crates/ruff_python_semantic/src/red_knot/types/infer.rs

```diff
@@ -0,0 +1,980 @@
+use std::sync::Arc;
+
+use rustc_hash::FxHashMap;
+
+use ruff_db::vfs::VfsFile;
+use ruff_index::IndexVec;
+use ruff_python_ast as ast;
+use ruff_python_ast::{ExprContext, TypeParams};
+
+use crate::module::resolver::resolve_module;
+use crate::module::ModuleName;
+use crate::name::Name;
+use crate::red_knot::semantic_index::ast_ids::{ScopeAstIdNode, ScopeExpressionId};
+use crate::red_knot::semantic_index::definition::{
+    Definition, ImportDefinition, ImportFromDefinition,
+};
+use crate::red_knot::semantic_index::symbol::{
+    FileScopeId, ScopeId, ScopeKind, ScopeSymbolId, SymbolTable,
+};
+use crate::red_knot::semantic_index::{symbol_table, ChildrenIter, SemanticIndex};
+use crate::red_knot::types::{
+    ClassType, FunctionType, IntersectionType, LocalClassTypeId, LocalFunctionTypeId,
+    LocalIntersectionTypeId, LocalUnionTypeId, ModuleType, Type, TypeId, TypingContext, UnionType,
+    UnionTypeBuilder,
+};
+use crate::Db;
+
+/// The inferred types for a single scope.
+#[derive(Debug, Eq, PartialEq, Default, Clone)]
+pub(crate) struct TypeInference {
+    /// The type of the module if the scope is a module scope.
+    module_type: Option<ModuleType>,
+
+    /// The types of the defined classes in this scope.
+    class_types: IndexVec<LocalClassTypeId, ClassType>,
+
+    /// The types of the defined functions in this scope.
+    function_types: IndexVec<LocalFunctionTypeId, FunctionType>,
+
+    union_types: IndexVec<LocalUnionTypeId, UnionType>,
+    intersection_types: IndexVec<LocalIntersectionTypeId, IntersectionType>,
```

Comment:
I'm not sure if it will be a lot more, or just some more. We won't duplicate types that we resolve from an other scope or module. We only duplicate the type if both scopes contain code paths joining the same two types. 

I also think that we can work around this with software engineers' favorite trick: one more layer of indirection. We could create a true `Interner` for types and create one per file and exclude the internet from the `Eq` implementation.